src/Rings/mpoly-graded.jl

```diff
@@ -1612,6 +1612,349 @@ function Base.show(io::IO, h::HilbertData)
   print(io, "Hilbert Series for $(h.I), data: $(h.data), weights: $(h.weights)")  ###new
 end
 
+########################################################################
+# Compute the Hilbert series from the monomial leading ideal using 
+# different bisection strategies along the lines of 
+# [Kreuzer, Robbiano: Computational Commutative Algebra 2, Springer]
+# Section 5.3.
+########################################################################
+exp_vec(f::MPolyElem) = first(exponent_vectors(f))
+
+_divides(a::Vector{Int}, b::Vector{Int}) = all(x->(x>=0), a-b)
+
+function _gcd(a::Vector{Vector{Int}}) 
+  length(a) == 1 && return a[1]
+  n = length(a[1]) # assumed to be the length of all entries of a
+  return [minimum([a[i][j] for i in 1:length(a)]) for j in 1:n]
+end
+
+function _are_pairwise_coprime(a::Vector{Vector{Int}})
+  length(a) <= 1 && return true
+  n = length(a)
+  for i in 1:n-1
+    for j in i+1:n
+      all(x->(x==0), _gcd([a[i], a[j]])) || return false
+    end
+  end
+  return true
+end
+
+### Assume that a is minimal. We divide by `pivot` and return 
+# a minimal set of generators for the resulting monomial ideal.
+function _divide_by(a::Vector{Vector{Int}}, pivot::Vector{Int})
+  shifted = [e-pivot for e in a]
+  # The good ones will contribute to a minimal generating 
+  # set of the lhs ideal.
+  good = [e for e in shifted if all(x->(x>=0), e)]
+
+  # The bad monomials come from those which hop over the boundaries of 
+  # the monomial diagram by the shift. Their span has a new 
+  # generator which is collected in `bad` a priori. It is checked 
+  # whether they become superfluous and if not, they are added to 
+  # the good ones.
+  bad = [e for e in shifted if !all(x->(x>=0), e)]
+
+  divides(a::Vector{Int}, b::Vector{Int}) = all(k->(k>=0), b - a)
```

Comment:
this computes (and thus allocates) the vector `b-a`, then runs over it to check the entries for being positive. I am guessing this will be quite a bit slower than doing the direct approach, i.e.,
```suggestion
  divides(a::Vector{Int}, b::Vector{Int}) = all(k->b[k]>=a[k], 1:length(a))
```
Indeed:
```
julia> v1 = collect(1:10) ; v2 = collect(1:10) * 2;

julia> @btime divides(v1,v2) ; @btime divides2(v1,v2);
  30.402 ns (1 allocation: 144 bytes)
  8.133 ns (0 allocations: 0 bytes)

julia> @btime divides(v2,v1) ; @btime divides2(v2,v1);
  26.800 ns (1 allocation: 144 bytes)
  4.333 ns (0 allocations: 0 bytes)
```

Reply:
Or `all((>=).(b, a))` for maximal julia style and compactness (but it is a bit slower, since unfortunately there is no "lazy" version).

Reply:
That's definite really nice and compact, but it performs worst of all the three options in my test (in Julia 1.8.1 on an M1 mac):
```
julia> using BenchmarkTools

julia> divides(a::Vector{Int}, b::Vector{Int}) = all(k->(k>=0), b - a);

julia> divides2(a::Vector{Int}, b::Vector{Int}) = all(k->b[k]>=a[k], 1:length(a));

julia> divides3(a::Vector{Int}, b::Vector{Int}) = all((>=).(b, a));

julia> divides4(a::Vector{Int}, b::Vector{Int}) = @inbounds all(k->b[k]>=a[k], 1:length(a));

julia> v1 = collect(1:10) ; v2 = collect(1:10) * 2;

julia> @btime divides(v1,v2) ; @btime divides2(v1,v2); @btime divides3(v1,v2);
  30.402 ns (1 allocation: 144 bytes)
  8.133 ns (0 allocations: 0 bytes)
  40.657 ns (2 allocations: 96 bytes)

julia> @btime divides(v2,v1) ; @btime divides2(v2,v1); @btime divides3(v2,v1);
  26.842 ns (1 allocation: 144 bytes)
  4.375 ns (0 allocations: 0 bytes)
  40.616 ns (2 allocations: 96 bytes)

julia> @btime divides(v1,v1) ; @btime divides2(v1,v1); @btime divides3(v1,v1);
  30.977 ns (1 allocation: 144 bytes)
  3.083 ns (0 allocations: 0 bytes)
  40.573 ns (2 allocations: 96 bytes)
```

BTW these numbers are from a "pure" Julia session, no Oscar loaded. For reasons I can't fathom, variants 1 and 3 are quite a bit slower with Oscar loaded, even though the machine code according to `@code_native` appears to be identical. Huh.